### ISSUE TITLE
Refine array types using `filter`

### ISSUE THREAD

**Issue by vkurchatkin** [facebook/flow#1414]
Not sure if it's possible, but it would be nice if flow could refine array type using `filter`:

``` js
function getArray(): Array<?{}> { [{}, {}, null, {}, null] }

const array: Array<{}> = getArray().filter(i => Boolean(i));
```

Simplest patterns that are already used to refine optional types would be just enough

**Comment by avikchaudhuri** [comment]
Yeah, this feature should eventually be supported, along with the ability to annotate functions like the one you pass to `filter` with "effects." It's on the TODO list, but not hi-pri right now.

**Comment by vkurchatkin** [comment]
Here is my proposal:

 - implement a magic type, that represents result of type check, for example `Typecheck<Foo>`;
 - `Typecheck` is subtype of boolean;
 - `Typecheck` is only allowed to be explicitly used as type of return value of a single-argument function:

```js
function isNumber(val: mixed): Typecheck<number> {
  if (typeof val === 'number') {
    return true;
  }

  return false;
}
```
 - if typecheck function returns `true`, that means that argument is of type `Foo`;
 - typecheck function is validated, so you can only return `true` if  `val` is of type `Foo` at this point;
 - typecheck function is only allowed to call other typecheck functions
 
For example, `Boolean` can be declared as `<T: {}>(v: ?T): Typecheck<T>`.

Finally, filter function could be declared like this:

```js
filter<T2>(fn: (v: T) => Typecheck<T2>): Array<T2>
```

**Comment by ctrlplusb** [comment]
I was just coming to post a similar issue.  I am experiencing the following:

<img width="596" alt="screenshot 2016-07-11 06 28 46" src="https://cloud.githubusercontent.com/assets/12164768/16720725/0ffde0ac-4731-11e6-8f0b-b3292d7fb417.png">

**Comment by ctrlplusb** [comment]
@gcanti mentions a possible workaround for now in #1915:

```typescript
const input: Array<number|string> = [1, 'a', 2, 3, 'b']
const result = ((input.filter(el => {
  return typeof el == 'number'
}): Array<any>): Array<number>)
```

**Comment by ctrlplusb** [comment]
Sorry for the spam, but another update on this.  Using `.reduce` seems to play much nicer with types.  Just annotate your accumulator and it's all good.

**Comment by gabro** [comment]
It seems that with more recent version of flow this is supported using a point-free style, e.g.

```js
declare function getArray(): Array<?{}>
const array: Array<{}> = getArray().filter(Boolean);
```

The snippet above type-checks on flow.

https://flowtype.org/try/#0CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMQMBBGGKATwAoBKALnjIsoB4B+AbwF8A+AKHjx+YPAGcM8WC0bMqrHr3gBeQsTk1aAOkRYIGEDGoAhHDgggoqWgG4gA

**Comment by vkurchatkin** [comment]
It doesn't work for generic predicates though

**Comment by gabro** [comment]
Right, that one is still missing.

**Comment by lll000111** [comment]
Any progress?

**Comment by bdrobinson** [comment]
+1 for this. In fact, since `?string` is just a specific case of a union, it'd be great more generally if Flow could filter any type union: ie, filtering redux actions by type:

```js
type SetNameAction = { type: "SET_NAME", name: string }
type SetAgeAction = { type: "SET_AGE", age: number }
type Action =  SetNameAction | SetAgeAction

const actions: Array<Action> = [] 

// currently this errors even though we can be certain that
// the array only contains `SetAgeAction`s
const setAgeActions: Array<SetAgeAction> = actions.filter(
    action => action.type === "SET_AGE")
```

Flow can already filter type unions with `if`/`switch` statements (which blew my mind the first time I saw it) – could a similar approach be applied to filter? (quite possibly not, I have no idea how these things work under the hood):

```js
const printActionName = (action: Action) => {
  
  // flow correctly errors because action could be a SetAgeAction
  const name: string = action.name
  
  if (action.type === "SET_NAME") {
    // no error here as flow correctly infers that action must be a
    // SetNameAction from the if statement
    const name: string = action.name
    console.log(name)
  }
}
```

Thanks – flow is awesome!

**Comment by bradennapier** [comment]
I run into quite a few situations where I am having to do value transformations whether it be filter or any other situation -- and while it isn't very difficult once you get the hang of it, I figured a little micro-package would be useful here to enforce type-safe value transformations in a standard manner.

Its late so I'll try to put together some nice examples that are actually useful soon -- for now there are a few documented and it should work very well.

It is published to npm as `flow-type-transformer` and is a runtime value transformer which is extremely tiny. 

Below is a completely useless example to illustrate the purpose.  it simply provides a nice little transform utility type & transformation class:

```js
import type { $Transform } from 'flow-type-transformer';
import createTransformer   from 'flow-type-transformer';

// pointless transformer example (since Flow can infer this)
const stringToNumber: $Transform<string, number> = createTransformer(str => Number(str))
const n = stringToNumber('1')
```

@ctrlplusb I would think something like this should make it far easier to build nice functional interfaces that remain type safe.  

https://github.com/Dash-OS/flow-type-transformer

**Comment by jcready** [comment]
@bradennapier this might pass flow type checking, but it will also throw a javascript runtime error because a `$call` property doesn't actually have any significance in javascript in spite of what flow thinks. Running this results in an error ([jsfiddle](https://jsfiddle.net/v0ttvjnm/)):

```js
class TypeTransformer {
  constructor(transformer) {
    this.transformer = transformer;
  }
  $call(v) {
    return this.transformer(v);
  }
}

function createTypeTransformer(fn) {
  return new TypeTransformer(fn);
}

function removeFalseyFromArray(v) {
  if (Array.isArray(v)) {
    return v.filter(el => el != null);
  }
  throw new TypeError('Value is not an Array');
}

const transform = createTypeTransformer(removeFalseyFromArray);

const before = ['one', undefined, 'two', null];

const after = transform(before); // Uncaught TypeError: transform is not a function
```

**Comment by jcready** [comment]
However, you can make your class work as intended in actual javascript by changing the class to return the transformer from the constructor:

```js
class TypeTransformer<Before, After> implements $Transform<Before, After> {
  transformer: $Transformer<Before, After>;
  constructor(transformer: $Transformer<Before, After>) {
    this.transformer = transformer;
    return transformer;
  }
  $call(v: Before): After {
    return this.transformer(v);
  }
}
```

This manages to trick flow into believing that calling `new TypeTransformer(foo)` returns an instance of `TypeTransformer`, but in actuality javascript is returning `foo` so now both flow and javascript are happy!

**Comment by bradennapier** [comment]
@jcready yes you are correct.  I had planned to change it to a standard function because at first the class was going to have things like `.filter` and `map` but then i had changed it to providing a custom transformer function.

However, your little workaround here is actually pretty perfect.  I have made the adjustment!  Thanks.

**Comment by bradennapier** [comment]
So this has been really powerful for us already.  For example, one thing that has always been a bit challenging is transforming object literal values.  

Say I have a response from an API that I need to format.  I want to make sure it fits the structure I require before passing it on.  

I want it to be an exact object type but it can't be enforced that way until I have normalized it otherwise you run into various issues when trying to modify it, etc. 

With the below we are able to retain 100% flowtype safety throughout the process:

```js
export const formatErrorResponse: $Transform<
  { ...Dash$ErrorResponse },
  $Exact<Dash$ErrorResponse>,
> = createTransformer(response => {
  const errors = response.errors || [];
  if (response && typeof response.errorMessage === 'string') {
    // If recieving the previous format for any reason,
    // convert to the new style array
    errors.push(response.errorMessage);
  }
  return {
    result: 'error',
    code: response.code || 501,
    uuid: response.uuid || '0',
    path: response.path,
    errors,
  };
});
```

**Comment by sibelius** [comment]
can we use `$Pred` and `$Refine` for this?

**Comment by Chaoste** [comment]
I also struggled with this and it is a bit frustating that this is open for two years. The tip by @ctrlplusb to use reduce was very helpful!

So instead of my old code:
```
items.filter(item => item.type === MY_TYPE)
```
I'm now using:
```
items.reduce(
    (arr, item) => (item.type === MY_TYPE ? [...arr, item] : arr),
    []
);
```

[Try Flow: forEach vs. reduce vs. filter](https://flow.org/try/#0C4TwDgpgBAggdiAogGwgWwnYUC8UDeAUFFKJAFxQDkMAcgJpUA0xUAhsMAE6UDO3ASzgBzFgF8A3IUJloAdQAWHAJJy2WACrhoeIiVmUqAZQAKiAMLKAYsvPNWwAcFR9BI8VMIBjAPZx+7Fw8sEFsIAA88EioGFhQAD5QiipqmtoAfLhQANr4pNqGphbWtsykTi7UAOwArAAcVFBiTAT5FNR0jC0c3IYALDUAbI1iALqe3n4BbEEAjJQwoRHJwKrqwFqQmXjZ44QzXAB0AGY+XIhsXgoAFAAeuJl6UALHUHeHsrg4eMZmljZ2ACUBFYJAOs0OYAArrwbrdAVISGJCGIEdJJv5sAcAEwLJbhFZrNJbLIHQ5cCAAEyhXgg11Y10uXhaXihQWBOEyrKCH20Xx+RX+pSgAH4cocJUyWWyuKMoJQpTlRoD0b5MYEuABmPFcMIEpSrVIbDKknnHATIYAQLh3B5QW68yD86iCkpAwhAA)

**Comment by virtue3** [comment]
> I also struggled with this and it is a bit frustating that this is open for two years. The tip by @ctrlplusb to use reduce was very helpful!
> 
> So instead of my old code:
> 
> ```
> items.filter(item => item.type === MY_TYPE)
> ```
> 
> I'm now using:
> 
> ```
> items.reduce(
>     (arr, item) => (item.type === MY_TYPE ? [...arr, item] : arr),
>     []
> );
> ```
> 
> [Try Flow: forEach vs. reduce vs. filter](https://flow.org/try/#0C4TwDgpgBAggdiAogGwgWwnYUC8UDeAUFFKJAFxQDkMAcgJpUA0xUAhsMAE6UDO3ASzgBzFgF8A3IUJloAdQAWHAJJy2WACrhoeIiVmUqAZQAKiAMLKAYsvPNWwAcFR9BI8VMIBjAPZx+7Fw8sEFsIAA88EioGFhQAD5QiipqmtoAfLhQANr4pNqGphbWtsykTi7UAOwArAAcVFBiTAT5FNR0jC0c3IYALDUAbI1iALqe3n4BbEEAjJQwoRHJwKrqwFqQmXjZ44QzXAB0AGY+XIhsXgoAFAAeuJl6UALHUHeHsrg4eMZmljZ2ACUBFYJAOs0OYAArrwbrdAVISGJCGIEdJJv5sAcAEwLJbhFZrNJbLIHQ5cCAAEyhXgg11Y10uXhaXihQWBOEyrKCH20Xx+RX+pSgAH4cocJUyWWyuKMoJQpTlRoD0b5MYEuABmPFcMIEpSrVIbDKknnHATIYAQLh3B5QW68yD86iCkpAwhAA)

Finally dealing with this myself and quite frustrated that I can't utilize filter very well.  I feel like a major tool in my belt is now missing :(

I just wanted to point out that your solution was an excellent starting point and Ive iterated on it in order to take out some extra object creations with the splat operator (albeit more/uglier code):

```js
items.reduce(
  (arr, item) => {
    if(item.type === MY_TYPE) {
      arr.push(item);
    }
   return arr;
 }
);
```

**Comment by Chaoste** [comment]
@virtue3 
Thanks for sharing your version!

However, I want to point out that the other workaround by @ctrlplusb ist also quiet handy. In my case I just needed to manually **redefine the return value** like this:
```
const result = (array.filter(MY_FILTER): Array<T>)
```

In my reducer, it was even easier to just **flow type the return value** of the superordinate function instead of using the above inline version:

```
const reducer = (state: MyArrayStateType = initialState, action: MyActionType): MyArrayStateType => {
    switch ...
        case ...
            return state.filter(MY_FILTER)
}
```

**Comment by choonkending** [comment]
Given how widely we use `filter`, is it worth having a discussion on whether there is room to change the type definition itself [Array's definition in Flow](https://github.com/facebook/flow/blob/v0.102.0/lib/core.js#L258)?

E.g.,  is something such as the following possible?

```js
// Naive implementation
filter<FilteredType: T>(predicate: (value: T, index: number) => boolean): Array<FilteredType>;
```

Here is a [Try Flow](https://flow.org/try/#0C4TwDgpgBAygrgIwCrggZwPIDMbAE4CWAdgOZQC8UA5AIJVQA+1AQlQNwBQHAxgPZFpgUAIYUoACmF48wkAC4oNabIA8gwqQB8ASgVKZIFfGSpMOfMRKaK1qQYB0WAgBtgEPOII2oX8n+p0jEy+-lSs2pwcwuIA2rRUADTUAMKJ1AAiVAC62kA) as a test case we wish to pass.

I'd love to not resort to `reduce` just to refine an array of values, as `filter` conveys a lot more in terms of readability in a large codebase rather than the powerful `reduce`. 

**tl;dr**

I'd like to know:
1. If refinement from `A -> SubType of A` is achievable with `filter` 
- not with the current definition of Array, but whether or not it is possible with Flow currently
2. What we can do to move this forward + if any help is needed from us/anyone in the community

**Comment by goodmind** [comment]
@choonkending 

1. $Pred and $Refine
2. Well, nothing, it's still experimental features

**Comment by FlavienBusseuil** [comment]
In addition to what @bdrobinson said, I guess we have the same situation with `find`:

```js
type Cat = { type: "cat" };
type Dog = { type: "dog" };
type Animal = Cat | Dog;

const cat: Cat = { type: "cat" };
const dog: Dog = { type: "dog" };
const animals: Array<Animal> = [cat, dog];

const aCat: ?Cat = animals.find(({ type }) => type === "cat");
// Cannot assign `animals.find(...)` to `aCat` because  string literal `dog` [1]
// is incompatible with  string literal `cat` [2] in property `type`.
```

**Comment by w01fgang** [comment]
This issue looks quite old, but, at least for me, it's still actual.
From the [definition](https://github.com/facebook/flow/blob/v0.102.0/lib/core.js#L257) of `.filter` it's clear that for now, the only way to filter is using `.filter(Boolean)`.
```javascript
filter(callbackfn: typeof Boolean): Array<$NonMaybeType<T>>;
```
Other methods are returning the same type as they receive.
```javascript
filter(callbackfn: (value: T, index: number, array: Array<T>) => any, thisArg?: any): Array<T>;
find(callbackfn: (value: T, index: number, array: Array<T>) => any, thisArg?: any): T | void;
```
It seems there is a way to filter without using chainig.
```javascript
type SubTypesOfString = 'A' | 'B';

declare function ABfilter<T, P: $Pred<1>>(v: Array<T>, cb: P): Array<$Refine<T,P,1>>;
function isValid(i): %checks {
  return i === 'A' || i === 'B';
}

const arr: Array<mixed> = ['A', 'C', 'D'];
const barr: Array<SubTypesOfString> = ABfilter(arr, isValid);
```

I found my way to make it work but for small arrays, because of additional iterations
```javascript
const arr: Array<mixed> = ['A', 'C', 'D'];
const a: Array<SubTypesOfString> = arr
  .map(el => is_string(el) ? el : null)
  .filter(Boolean);
```
for the @FlavienBusseuil's example
```javascript
const aCat: ?Cat = animals
  .map((el) => el.type === "cat" ? el : null)
  .filter(Boolean)
  .find(({ type }) => type === "cat");
```
[try](https://flow.org/try/#0C4TwDgpgBAygrgIwCrggZwPIDMbAE4CWAdgOZQC8UA5AIJVQA+1AQlQNwBQHAJhAMYAbAIZ5oWOET7ACAeyJQAtiAD6WAgOAQ8AHiQAaKAAUAXFAAkh0d20BGAHx2AFADdTNPHiEhddg3wSmhgCUbh5e2mYAShBqRBC6eoZ69nacHM4iUCJ4oZ7eCgQAHhDcdhRQANq0VAZUAMI11AAiVAC6nHxyaMBQCNnlSqrqmniO2QYEaMrdhKRBnI59Hrnh8MiomDj4xCR281ziktJyUJPT26SOBCFQAKR8ABb8ANZoUADeHFBQosBwePICBRyJRqowmECQaDWJwAL5cTpEbpZFbeNYoSCbXCzXblbIAOgUQjAjggAgoZTOMx2pIEQSgAH4oGSoKYiHABHT8WoNFpHMwZDIBBAhER5kA)

**Comment by FlavienBusseuil** [comment]
@w01fgang thanks for your help.

Still I'm concerned by the boilerplates and performance issues added by those two loops.

Also I came up with another issue. Filtering based on object attribute presence:

```js
type Cat = { meow: true };
type Dog = { woof: true };
type Animal = Cat | Dog;

const cat: Cat = { meow: true };
const dog: Dog = { woof: true };
const animals: Array<Animal> = [cat, dog];

const cats1: Array<Cat> = animals.filter((animal) => animal.meow);
// Cannot assign `animals.filter(...)` to `cats1` because property `meow` is 
// missing in `Dog` [1] but exists in `Cat` [2] in array element.

const cats2: Array<Cat> = animals
	.map((animal) => animal.meow ? animal : null)
	// Cannot assign `animals.map(...).filter(...)` to `cats2` because property 
	// `meow` is missing in `Dog` [1] but exists in `Cat` [2] in array element.
	.filter(Boolean);

const cats3: Array<Cat> = animals
	.reduce((cats, animal) => animal.meow ? [...cats, animal] : cats, []);
	// Cannot assign `animals.reduce(...)` to `cats3` because property `meow` is 
	// missing in `Dog` [1] but exists in `Cat` [2] in array element.
```

I really don't get this one... why `animal.meow ?` do not distinguish which type of object I'm working with? Am I missing something? 🤔

**Comment by noppa** [comment]
@FlavienBusseuil That problem goes away if you make the types [exact](https://flow.org/en/docs/types/objects/#toc-exact-object-types).
```javascript
type Cat = {| meow: true |};
type Dog = {| woof: true |};
```
without exact types, `Dog` could also have a `meow` property of some type, because extra properties are allowed, and the check would erroneously mark it as a `Cat`.

Maybe a helper function [like this](https://flow.org/try/#0PYBwhgjgrgpgBAFwJ4ngZQNYEsRwLxwDOSAtgEbAA2AUAMbAB2hCR2uBapFlAFAOQAzLJQQwATgBUUMAPSE2fAJTVqAqA1oIsjOEJHipqADwAxMcBKGYAGjgTgVgHw9qcOJSzMAXHAAkAJRgwABMAeQZKJABBMTEwJFNzS2lHa1ddBh8eADcwSh8zCytbeRwfTBxFfEc4CtwAHzsHaWpFH3srAG0AXTgAb3T6JhYxGEIoEXbm1B78OB70gWAxOB4h5jgsURI4YAF3TwQqgbc3dZG5gQYeLZgSErZlU839nhWAQjwCUpBj9OfRuMRAA6EBQQgACx4bx8YAYSEUT1OAF90qi3KMEFAxAw4ICJghqKi6IwNgwoCRCAA5NAIMRYBgAc0Ic06AEZbAAmWx8ADMfG6JOGcHJlMuwlEkmkPFF1Np9KZhFsOTyD0q1UQ0j2cFylHwXzgfFFZHEfDgAH4dXk4D4fspqDIZHATJRgAB3ODqYLiZhw4IshAQsAsLZ8FlgEUUk0rMCxeIi93UWWdAAM3WBCGAJiwAA8YMEeJzlMm0xms7n8wWi0A) would help a bit with the loop cost and boilerplate.

**Comment by FlavienBusseuil** [comment]
@noppa thanks! As I moved with the new syntaxe without the pipes recently I totally forgot about them!

**Comment by w01fgang** [comment]
@noppa I found that there is a cast `any` and it doesn't work without it. In my codebase, I use `flow strict` preventing `any` types. I changed a bit your code replacing symbol by null, and it works! Thanks!
[try](https://flow.org/try/#0GYVwdgxgLglg9mABMGAbKBTATgFQJ4AOGAPAGJZwC2+RANIjnDRgHwAUAUIoqjAM5QAXIgAkAJQwBDACYB5MKjwBBLFkl4yFaoVa0uyMMLYA3SamHkqzAJSIAvCwZMdiAD6IwIVKg7XhjZgBtAF1EAG99CAQBRCwMPi8hJyDQu0QQ-WA4LEQ2KLAYmExKRDhgHn4oWwjubnyYnLTgMDYijEprfW4YcrYcgEI7NM9vaq7auIT0ADoCED4ACz7O2sQAX30N7jioECwkScSODY56qA8QSj4AOQBlKCwYMABzPnt0gEZ6ACZ6AHIAMx-YKnaLnTxXd4odDYZhsCE3e6PF58egmMy2ByIKA6MqIUyoexDRB-CEAI2wf0QAH58WZEMIRqhOhwAPSsxCkVBwADuiHA0mwAkkYGkbygC0k5yKfzekgulApOUkqnUHl5HARgQADMFplA4KQYAAPDDSNjfTrsxAAUWNRGgiGwFCwmsufB1eoNRtN0nNlqAA)

**Comment by FezVrasta** [comment]
I think I'm missing something, why is `Array.prototype.filter` so limited? `Array.prototype.reduce` works perfectly and it's basically a superset of `filter`, why can't `filter` be defined as a constrained `reduce`?

example:

```js
Array.prototype.filter = (callback) => this.reduce((a, b) => {
  if (callback(b) {
    return a.concat(b);
  } else {
    return a;
  }
}, []);
```

**Comment by nnmrts** [comment]
@FezVrasta You're not missing anything. This has been a pattern for 2 years now. As @goodmind wonderfully pointed out when asked about "What we can do to move this forward + if any help is needed from us/anyone in the community" by @choonkending:

> Well, nothing, it's still experimental features

I don't really know what "experimental" in this context means, but my guess is language-wise, which would be just plain wrong in the case of `filter`, [which has been added to Javascript before Windows Vista was released](https://developer.mozilla.org/en-US/docs/Archive/Web/JavaScript/New_in_JavaScript/1.6).

But that wouldn't be the first time flow just didn't support a 15-year-old, core part of the ES standard because "it's unsound" or something like that. But actually, @goodmind has a point, you **really** shouldn't try do anything here, unless you want to waste your time.

But you do you, good luck and have fun! 💪

**Comment by noppa** [comment]
@nnmrts I think @goodmind was referring the experimental Flow features `$Pred` and `$Refine` there, that could've possibly been used to impement this feature in userland.

Based on [this thread](https://github.com/facebook/flow/issues/2464#issuecomment-471114363), though, those features are not going to become available for general use.

**Comment by FezVrasta** [comment]
Is there any way to override the built-in definition without copying the whole built-in Flow definitions?

**Comment by lyleunderwood** [comment]
@FezVrasta see https://github.com/facebook/flow/pull/7919

**Comment by andrewmclagan** [comment]
Filtering arrays based on types is such fundamental functionality, it leads one to believe development of Flow is basically dead. R.I.P Flow.

**Comment by w01fgang** [comment]
Well, not really. It's just stupid typecasting.
<img width="610" alt="Screenshot 2022-04-12 at 10 09 32" src="https://user-images.githubusercontent.com/8686751/162924540-d0df7f46-898c-4c81-a3bd-2906084212fe.png">
> Filtering arrays based on types is such fundamental functionality, it leads one to believe development of Flow is basically dead. R.I.P Flow.

Also, your original message is quite offensive. TS is hardly pushed by MS while people love Flow and decide to use it because of its value.

**Comment by SamChou19815** [comment]
This now works with type guards
